### PR TITLE
CORTX-34050: Apply limit_time filter on .gz files.

### PR DESCRIPTION
**Pull request by shriya-deshmukh**
# Problem Statement
- Handle processing for .gz files while applying the limit_time filter.

# Design
-  For Bug describe the fix here. 
-  For Feature, Post the link to the solution page on the confluence CORTX Foundation Library 

# Coding 
-  Coding conventions are followed and code is consistent [Y/N]:  Y
-  Confirm All CODACY errors are resolved [Y/N]:  Y

# Testing
- Are test cases updated along with code changes due to Enhancements/Bugs [Y/N]:
- Confirm that new test cases are added to regression and sanity plan files and relevant feature plan files [Y/N]:
- Confirm that Test Cases are added for new features added [Y/N]: 
- Confirm Test Cases cover Happy Path, Non-Happy Path and Scalability [Y/N]: 
- Confirm Testing was performed with installed RPM/K8s deployment [Y/N]:  Y

# Review Checklist 
####  Before posting the PR please ensure:
- [x] PR is self reviewed
- [ ] Is there a change in filename/package/module or signature [Y/N]: 
- [ ] If yes for above point, Is a notification sent to all other cortx components [Y/N]
- [ ] New package/s added to setup.py?
- [x] Jira is updated
- [x] Check if the description is clear and explained. 
- [ ] Check Acceptance Criterion is defined. 
- [ ] All the tests performed should be mentioned before Resolving a JIRA. 
- [ ] Verification needs to be done before marked as Closed/Verified.

# KVstore/Confstore feature (changes/fixes) checklist
#### Confirm changes are made for:
- [ ] ConfStore
- [ ] KVStore
- [ ] ConfCli
- [ ] Test cases added for all above 3
- [ ] changes done in all the KVpayloads (ConsulKvPayload, IniKvPayload, KvPayload)

# Documentation
- [ ] Changes done to WIKI / Confluence page

### CODE REVIEW

**Review by sachitanands — APPROVED** [review]
looks good.

File: py-utils/src/utils/support_framework/log_filters.py [hunk 1]
```diff
@@ -175,45 +177,92 @@ def limit_time(src_dir, dest_dir, duration, file_name_reg_ex,
...
+        list_of_files = filter(lambda f: os.path.isfile(os.path.join(src_dir, f)),
+            os.listdir(src_dir))
+        # sort the files based on last modification time in descending order
+        list_of_files = sorted(list_of_files,
+            key=lambda f: os.path.getmtime(os.path.join(src_dir, f)),
+            reverse=True)
+        for file in list_of_files:
+            file_extension = pathlib.Path(file).suffix
+            # Ignore processing of other file format.
+            if file_extension not in supported_file_types:
```
Comment (cdeshmukh) [inline py-utils/src/utils/support_framework/log_filters.py:196]: log.warn here, stating {filename} being skipped

File: py-utils/src/utils/support_framework/log_filters.py [hunk 2]
```diff
@@ -175,45 +177,92 @@ def limit_time(src_dir, dest_dir, duration, file_name_reg_ex,
...
-                            # to capture such logs, we have set a flag
-                            # include_lines_without_timestamp = True
-                            except ValueError:
-                                if include_lines_without_timestamp:
-                                    fd_out.write(line)
-                        line = fd_in.readline()
-                try:
+                if file.endswith('.gz'):
+                    # Reading gzip file content without extracting
+                    with gzip.open(in_file, 'rt') as fd_in:
```
Comment (cdeshmukh) [inline py-utils/src/utils/support_framework/log_filters.py:203]: gzip allows to read directly gzipped file, without the need to deflate it first ?

File: py-utils/src/utils/support_framework/log_filters.py [hunk 3]
```diff
@@ -175,45 +177,92 @@ def limit_time(src_dir, dest_dir, duration, file_name_reg_ex,
...
+                        lines = fd_in.readlines()
+                else:
+                    with open(in_file, 'r') as fd_in:
+                        lines = fd_in.readlines()
```
Comment (cdeshmukh) [inline py-utils/src/utils/support_framework/log_filters.py:207]: >  lines = fd_in.readlines()

above line is common, can be pulled out of if-else

File: py-utils/src/utils/support_framework/log_filters.py [hunk 4]
```diff
@@ -175,45 +177,92 @@ def limit_time(src_dir, dest_dir, duration, file_name_reg_ex,
...
+            if not lines:
+                continue
+            if file.endswith('.gz'):
```
Comment (cdeshmukh) [inline py-utils/src/utils/support_framework/log_filters.py:210]: why this gz check again? it was done on line 201

File: py-utils/src/utils/support_framework/log_filters.py [hunk 5]
```diff
@@ -175,45 +177,92 @@ def limit_time(src_dir, dest_dir, duration, file_name_reg_ex,
...
+        is_log_written_to_file = False
+        start_time, end_time = FilterLog._parse_duration(duration)
+        for line in contents:
+            regex_res = re.search(timestamp_regex, line)
+            log_duration = regex_res.group(0) if regex_res else None
+            if log_duration:
+                # convert log timestamp to datetime object
+                try:
+                    log_time = datetime.strptime(log_duration, datetime_format)
+                    if start_time <= log_time and log_time <= end_time:
```
Comment (cdeshmukh) [inline py-utils/src/utils/support_framework/log_filters.py:247]: Shouldn't it be 
 if start_time >= log_time

File: py-utils/src/utils/support_framework/log_filters.py (hunk 2 above)
Comment (shriya-deshmukh) [reply py-utils/src/utils/support_framework/log_filters.py:203]: Yes, here we are reading .gz file in text mode.

File: py-utils/src/utils/support_framework/log_filters.py (hunk 3 above)
Comment (shriya-deshmukh) [reply py-utils/src/utils/support_framework/log_filters.py:207]: Yes, but we have defined it under with open statement.
We can use simple open() also to open the file and move this line outside if-else as below. 
example:
```
if file.endswith('.gz'):
      # Reading .gz file content without extracting.
       fd_in = gzip.open(in_file, 'rt')
else:
      fd_in = open(in_file, 'r')
lines = fd_in.readlines()
fd_in.close()
```
But the best way to open/handle the file in python is - 'with open' as it also provides a way for cleanup. Hence I did the code change this way.


Should I change it as per the above example code or should I keep as it is?

File: py-utils/src/utils/support_framework/log_filters.py (hunk 4 above)
Comment (shriya-deshmukh) [reply py-utils/src/utils/support_framework/log_filters.py:210]: Here we are opening op_file the file in append mode. Previously we check this to open src_file file in reading mode.

File: py-utils/src/utils/support_framework/log_filters.py (hunk 5 above)
Comment (shriya-deshmukh) [reply py-utils/src/utils/support_framework/log_filters.py:247]: No, We need the logs which are generated between the log duration - ie. - start_time and end_time.
Example: if log_line is generated at 30-08-2022 2:46:00.
and start_time = 30-08-2022 4:00:00 and end_time = 30-08-2022 5:00:00.
So in that case we should not add log_line in our final result.

File: py-utils/src/utils/support_framework/log_filters.py (hunk 1 above)
Comment (shriya-deshmukh) [reply py-utils/src/utils/support_framework/log_filters.py:197]: Done

File: py-utils/src/utils/support_framework/log_filters.py [hunk 6]
```diff
@@ -175,45 +177,92 @@ def limit_time(src_dir, dest_dir, duration, file_name_reg_ex,
...
+                        include_lines_without_timestamp = True
+                        is_log_written_to_file = True
+                        if encode:
+                            line = line.encode('utf-8')
+                        fd_out.write(line)
+                    elif log_time > end_time and include_lines_without_timestamp:
+                        include_lines_without_timestamp = False
+                        is_log_exceeded = True
```
[the remark this answers is not in the document]
Comment (shriya-deshmukh) [reply py-utils/src/utils/support_framework/log_filters.py]: Done

File: py-utils/src/utils/support_framework/log_filters.py (hunk 4 above)
Comment (cdeshmukh) [reply py-utils/src/utils/support_framework/log_filters.py:212]: oh ok, missed that

File: py-utils/src/utils/support_framework/log_filters.py (hunk 3 above)
Comment (cdeshmukh) [reply py-utils/src/utils/support_framework/log_filters.py:209]: sure, lets keep it same way then

File: py-utils/src/utils/support_framework/log_filters.py (hunk 5 above)
Comment (cdeshmukh) [reply py-utils/src/utils/support_framework/log_filters.py:249]: got it, sorry for the mis-understanding

**Review by cdeshmukh — APPROVED** [review]

**Pull request merged**